src/test/net_tests.cpp

```diff
@@ -771,4 +773,128 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
+}
+
+// Returns true if any of the node ids in node_ids are selected for eviction.
+bool IsEvicted(std::vector<NodeEvictionCandidate> candidates, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    Shuffle(candidates.begin(), candidates.end(), random_context);
+    const Optional<NodeId> evicted_node_id = SelectNodeToEvict(std::move(candidates));
+    if (!evicted_node_id) {
+        return false;
+    }
+    return std::find(node_ids.begin(), node_ids.end(), *evicted_node_id) != node_ids.end();
+}
+
+// Create number_of_nodes random nodes, apply setup function candidate_setup_fn,
+// apply eviction logic and then return true if any of the node ids in node_ids
+// are selected for eviction.
+bool IsEvicted(const int number_of_nodes, std::function<void(NodeEvictionCandidate&)> candidate_setup_fn, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates = GetRandomNodeEvictionCandidates(number_of_nodes, random_context);
+    for (size_t i = 0; i < candidates.size(); ++i) {
+        candidate_setup_fn(candidates[i]);
+    }
+    return IsEvicted(candidates, node_ids, random_context);
+}
+
+namespace {
+// The number of eviction candidates at (or above) which we are guaranteed that
+// the current eviction logic will find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+constexpr int GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE{29};
+
+// The number of eviction candidates at (or below) which we are guaranteed that
+// the current eviction logic won't find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+constexpr int GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW{20};
+
+constexpr int NODE_EVICTION_TEST_ROUNDS{10};
+
+constexpr int NODE_EVICTION_TEST_UP_TO_N_NODES{200};
+} // namespace
+
+static_assert(GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE > GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW);
+
+BOOST_AUTO_TEST_CASE(node_eviction_test)
+{
+    FastRandomContext random_context{true};
+
+    for (int i = 0; i < NODE_EVICTION_TEST_ROUNDS; ++i) {
+        for (int number_of_nodes = 0; number_of_nodes < NODE_EVICTION_TEST_UP_TO_N_NODES; ++number_of_nodes) {
+            // Verify correctness of GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE
+            // and GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW.
+            if (number_of_nodes <= GUARANTEED_NON_EVICTION_AT_N_CANDIDATES_OR_BELOW) {
+                BOOST_CHECK(!SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));
+            }
+            if (number_of_nodes >= GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE) {
+                BOOST_CHECK(SelectNodeToEvict(GetRandomNodeEvictionCandidates(number_of_nodes, random_context)));
+            } else {
+                continue;
```

Comment:
I think you're not running the following checks if the number of nodes is between the min and the max because of this `continue`. Is that the intention?

Reply:
I don't get this either

Reply:
I've proposed an alternative that might be easier to parse.
https://github.com/bitcoin/bitcoin/pull/20477/files#r539574518

Reply:
It's "don't run the checks if the number of nodes is less than the max".  But, I'm not sure why we would do this...

Reply:
@narula Oh, good catch! The `continue` should be dropped. I don't remember what the apparently incorrect reasoning behind it was TBH :)

Thanks for catching it!